smart/model.py

```diff
@@ -2290,3 +2274,42 @@ def adjust_dt(self):
         #     dt_scale = min(dt_scale * 0.8, 0.8)
         dt_cur = float(self.dt) * dt_scale
         self.set_dt(dt_cur)
+
+    def create_restriction(self, u, mesh_function, value):
+        """
+        Restrict a function on a submesh to a subset of parent entities
+        (same dimension as the submesh)
+
+        :param u: Function on submesh
+        :param mesh_function: MeshFunction marking the subset of parent entities
+                            (same dimension as the cells of the submesh)
+        :param value: Value in MeshFunction marking the subset of parent entities
+        :return: New restricted function
+        """
+        submesh = u.function_space().mesh()
+
+        # Compute local cells in submesh marked by parent meshtag
+        mesh_ref = self.parent_mesh.dolfin_mesh
+        sub_to_parent_map = submesh.topology().mapping()[mesh_ref.id()].cell_map()
+        marked_sub_entities = mesh_function.array()[sub_to_parent_map] == value
+        local_indices = np.flatnonzero(marked_sub_entities)
+
+        # Find all degrees of freedom to transfer data from
+        V = u.function_space()
+        u_new = d.Function(u.function_space())
+        vector = u_new.vector()
+        dof_list = [V.dofmap().cell_dofs(cell) for cell in local_indices]
+        if len(dof_list) == 0:
+            transfer_dofs = np.array([])
+        else:
+            transfer_dofs = np.unique(np.hstack(dof_list))
+        im = V.dofmap().index_map()
+        num_local = im.local_range()[1] - im.local_range()[0] + 1
```

Comment:
Added one here as a correction, but let me know if I was incorrect.

Reply:
I think that is incorrect, as the range is [a,b), say [0,N) meaning N elements

Reply:
I thought the range should include N in this case (it looked like the number of dofs associated with running on a single process was in fact equal to `im.local_range()[1] - im.local_range()[0] + 1`. But I can double check tomorrow.

Reply:
I changed this back, I just had confused myself, but you were right :)